### ISSUE TITLE
SSL issues again

### ISSUE THREAD

**Issue by JenoraFeuer** [fchat-pidgin/fchat-pidgin#173]
So, sometime between November 10th when everything worked, and November 17th when I tried again, the F-list plugin started failing to connect, claiming 'SSL Handshake Failed'.  This is with the latest git download of this plugin, and pidgin 2.13.0.

The debug log didn't give me any more information:
(16:30:01) proxy: Connecting to www.f-list.net:443.
(16:30:01) proxy: Connected to www.f-list.net:443.
(16:30:01) gnutls: Starting handshake with www.f-list.net
(16:30:02) gnutls: Handshake failed. Error A TLS fatal alert has been received.
(16:30:02) flist: Web Request failed with error message: Unable to connect to www.f-list.net: SSL Handshake Failed
(16:30:02) flist: Failed to obtain the global list of profile fields. Error Message: Unable to connect to www.f-list.net: SSL Handshake Failed

The connection to fchat-pidgin.github.io worked without a problem.

Note that this happened not only on both Pidgin and Finch on my Linux box, but also Pidgin on my Windows laptop, so I don't think it's just a gnutls issue.  (And since it's gnutls, the NSS fix is unused.)

A wireshark trace showed much the same, a Client Hello, followed by an Alert (Level: Fatal, Description: Handshake Failure).

Firefox, of course, can connect just fine, and I have a Wireshark trace of that as well.

The one thing that stood out to me as different between the Pidgin attempt at connecting to f-list and the Firefox connection was some of the extensions, particularly Firefox is using the SNI (Server Name Indication) extension while Pidgin isn't.  Pidgin also isn't using SNI when connecting to fchat-pidgin.github.io, so I don't know if that's the real issue or not.  And if this is the problem, I don't know whether it can be fixed from the plugin or not.

The Discord and Telegram plugins both work, however.

Anybody else seeing this?

**Comment by JenoraFeuer** [comment]
Just adding another comment: the F-Chat 3.0 app downloaded from https://wiki.f-list.net/F-Chat_3.0/Installation seems to successfully make the SSL connection.  It's also using SNI in the Wireshark trace.

**Comment by JenoraFeuer** [comment]
And it looks like my theory was correct.  I added the following code to pidgin-2.13.0/libpurple/plugins/ssl/ssl-gnutls.c at line 409, after the handshake_handler was installed, but before the initialization of a 0 timeout to kick off the handshake callback:

        if (gsc->host) {
                int ret = gnutls_server_name_set(gnutls_data->session, GNUTLS_NAME_DNS, gsc->host, strlen(gsc->host)+1);
                if (ret == GNUTLS_E_SUCCESS) {
                        purple_debug_info("gnutls", "Set SNI to %s\n", gsc->host);
                } else {
                        purple_debug_info("gnutls", "Set SNI error %d\n", ret);
                }
        }
Unless it literally was changed on the f-list end in the hour or less I spent messing around with this after the previous failure, SNI support is the issue.

Unfortunately, as far as I can tell this means it needs to be fixed within Pidgin itself, because the fchat plugin doesn't really have a chance to splice itself in before the SSL handshake happens, especially when it's getting the kinks list or the like using purple_util_fetch_url_request() which doesn't exactly have a pre-handshake callback point.

**Comment by JenoraFeuer** [comment]
And there was already a ticket for it twenty months ago, which is still listed as 'unassigned' and which currently isn't on either the 2.14.0 or 3.0.0 milestone lists: https://developer.pidgin.im/ticket/17300

**Comment by vobe** [comment]
Thanks for investigating this, it looks like an GnuTLS specific issue. But did you say it also affects Windows ? The windows version should use NSS. 

You can try commenting on the pidgin issue, the fact that Cloudflare requires this change now might convince them to merge it, since Cloudflare is pretty much fronting the whole internet nowadays...

Until it's fixed on the Pidgin side, looks like you'll have to use NSS. Which still works fine, I'm using it everyday.

**Comment by JenoraFeuer** [comment]
Well, it affected my Windows laptop, but aside from a quick check, I didn't do any investigation there.

(SNI jumped out at me because I've had to deal with it at my day job; its purpose was to allow a single site to manage separate certificates for virtual hosts, but it's also used in a number of firewalls these days, because the firewall can't do filtering by site name on an encrypted connection any other way.  So I had to add SNI support to the old websocket code we were using.)

I run Gentoo, so everything gets built on-site anyway; the person who made the build package seems to have assumed that everybody would want only gnutls or nss, and it only builds one or the other.  I had to enable gnutls for something else, so it's defaulting to that.  I might be able to convince the Gentoo support to add this patch; they'd probably add it before Pidgin would.  And until then I can use the patch I just built.

Looks like I'll have to register to the Pidgin Trac setup.  Ah well, I was considering starting work on building my own plugin anyway, this is as good a time as any to start.
